tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/llm/TestWeightedSum.java

```diff
@@ -0,0 +1,359 @@
+        final int seqLen = 16;
+        final int pos = 15;
+        final int headSize = 32;
+        final int localSize = 16;
+        final int kvDim = headSize * numHeads;
+        final int kvMul = 1;
+        final int loff = 0;
+
+        // Allocate data arrays with known values
+        FloatArray attScores = new FloatArray(numHeads * seqLen);
+        FloatArray valueCache = new FloatArray(seqLen * kvDim);
+        FloatArray output = new FloatArray(numHeads * headSize);
+        FloatArray expectedOutput = new FloatArray(numHeads * headSize);
+        FloatArray debugData = new FloatArray(numHeads * localSize * 5); // Additional debug information
+
+        // Initialize with controlled values
+        Random random = new Random(42);
+
+        // Print head 0, dimension 0 values for clarity
+        System.out.println("Initializing test data:");
+
+        // Initialize attention scores - use softmax-normalized values
+        // Each head will have attention focused on a different position
+        for (int h = 0; h < numHeads; h++) {
+            int attOffset = h * seqLen;
+            float sum = 0.0f;
+
+            // Generate random values and calculate their sum
+            for (int t = 0; t <= pos; t++) {
+                float value = random.nextFloat();
+                attScores.set(attOffset + t, value);
+                sum += value;
+            }
+
+            // Normalize to create a proper probability distribution
+            for (int t = 0; t <= pos; t++) {
+                attScores.set(attOffset + t, attScores.get(attOffset + t) / sum);
+            }
+        }
+
+        // Initialize value cache with controlled values
+        for (int t = 0; t <= pos; t++) {
+            for (int h = 0; h < numHeads; h++) {
+                for (int i = 0; i < headSize; i++) {
+                    int valueOffset = loff + t * kvDim + (h / kvMul) * headSize + i;
+                    float value = random.nextFloat() * 2 - 1; // Values between -1 and 1
+                    valueCache.set(valueOffset, value);
+                }
+            }
+        }
+
+        // Do a special case for debugging - make head 0, dimension 0 values easier to track
+        // Set first value in first sequence position to a specific value for tracking
+        valueCache.set(loff + 0 * kvDim + 0 * headSize + 0, 1.0f);
+
+        // Print attention weights for head 0
+        System.out.println("Attention weights for head 0:");
+        for (int t = 0; t <= pos; t++) {
+            System.out.printf("Position %d: %.6f%n", t, attScores.get(0 * seqLen + t));
+        }
+
+        // Print the first few value vectors for head 0
+        System.out.println("First few values for head 0, dimension 0:");
+        for (int t = 0; t <= Math.min(pos, 3); t++) {
+            int valueOffset = loff + t * kvDim + (0 / kvMul) * headSize + 0;
+            System.out.printf("Position %d: %.6f%n", t, valueCache.get(valueOffset));
+        }
```

Comment:
Are this print statements valuable for when we run the unit-tests? or are they added for documentation purposes? If it is for documentation, perhaps we can add one flag that is FALSE by default, in order to reduce the logged output when we run all unit-tests. 

In my opinion, when we run the tests we should focus on PASS/FAIL results. What do you think?